File: padelizou-arcade/js/arbitro.js
/* Árbitro: transforma os eventos da física em decisão de ponto, seguindo as regras do padel.
 *
 * - Depois de um golpe, a bola precisa cruzar a rede e quicar no chão do outro lado ANTES de
 *   tocar qualquer parede de lá. Parede direto = ponto pra quem recebia.
 * - Quicou do lado de quem bateu (não passou, ou parou na rede) = ponto pra quem recebia.
 * - Bater nas próprias paredes antes de cruzar é permitido.
 * - Depois de quicar, a bola pode bater nas paredes à vontade; o segundo quique encerra o
 *   ponto a favor de quem bateu. Se voltar pela rede depois de quicar, idem.
 * - Saiu por cima da parede: ponto pra quem bateu se já tinha quicado; senão, contra.
 * - Saque: precisa quicar na caixa diagonal; erro é falta, duas faltas é ponto; tocar a
 *   rede e cair na caixa é let (repete). Quem recebe não pode voleiar o saque.
 * - O mesmo time não bate duas vezes seguidas. */
(function () {
  'use strict';
  const raiz = typeof window !== 'undefined' ? window : globalThis;
  const Padel = raiz.Padel || (raiz.Padel = {});
  const Quadra = Padel.Quadra || (typeof require === 'function' ? require('./quadra.js') : null);

  class Arbitro {
    constructor() {
      this.faltas = 0;
      this._zerarRally();
    }

    _zerarRally() {
      this.emSaque = false;
      this.golpeador = null;         // time que bateu por último
      this.quicouNoReceptor = false;
      this.tocouARede = false;
      this.cruzou = false;
      this.caixa = null;
    }

    novoPonto() { this.faltas = 0; this._zerarRally(); }

    iniciarSaque(time, caixa) {
      this._zerarRally();
      this.emSaque = true;
      this.golpeador = time;
      this.caixa = caixa;
    }

    registrarGolpe(time) {
      this._zerarRally();
      this.golpeador = time;
    }

    get receptor() { return this.golpeador === null ? null : 1 - this.golpeador; }

    podeGolpear(time) {
      if (this.golpeador === null || this.golpeador === time) return false;
      if (this.emSaque && !this.quicouNoReceptor) return false;
      return true;
    }

    // Devolve null (segue), { tipo: 'ponto', para, motivo }, { tipo: 'falta', motivo } ou { tipo: 'let' }.
    processar(evento) {
      if (this.golpeador === null) return null;
      const ladoDoGolpeador = Quadra.ladoDoTime(this.golpeador);
      switch (evento.tipo) {
        case 'cruzouRede':
          if (evento.para !== ladoDoGolpeador) { this.cruzou = true; return null; }
          if (this.quicouNoReceptor) return this._ponto(this.golpeador, 'voltouPeloVidro');
          return null;
        case 'rede':
          this.tocouARede = true;
          return null;
        case 'quique':
          if (evento.lado === ladoDoGolpeador) return this._contraOGolpeador(this.tocouARede ? 'rede' : 'naoPassou');
          if (this.quicouNoReceptor) return this._ponto(this.golpeador, 'doisQuiques');
          this.quicouNoReceptor = true;
          if (this.emSaque) {
            if (!Quadra.dentroDaCaixa(evento.x, evento.y, this.caixa)) return this._contraOGolpeador('foraDaCaixa');
            if (this.tocouARede) return { tipo: 'let' };
          }
          return null;
        case 'parede':
          if (evento.lado === ladoDoGolpeador) return this.emSaque ? this._contraOGolpeador('paredePropria') : null;
          if (!this.quicouNoReceptor) return this._contraOGolpeador('paredeSemQuicar');
          return null;
        case 'saiu':
          if (this.quicouNoReceptor) return this._ponto(this.golpeador, 'fora');
          return this._contraOGolpeador('fora');
        default:
          return null;
      }
    }

    _contraOGolpeador(motivo) {
      if (!this.emSaque) return this._ponto(this.receptor, motivo);
      this.faltas += 1;
      if (this.faltas >= 2) return this._ponto(this.receptor, 'duplaFalta');
      return { tipo: 'falta', motivo };
    }

    _ponto(para, motivo) {
      return { tipo: 'ponto', para, motivo };
    }
  }

  Padel.Arbitro = Arbitro;
  if (typeof module !== 'undefined' && module.exports) module.exports = Arbitro;
})();

File: padelizou-arena/Padel.Core/Arbitro.cs
namespace Padel.Core;

public enum Motivo { DoisQuiques, Rede, NaoPassou, ParedeSemQuicar, ParedePropria, Fora, VoltouPeloVidro, DuplaFalta, ForaDaCaixa, BolaMorta }
public enum TipoDeDecisao { Ponto, Falta, Let }

public readonly record struct Decisao(TipoDeDecisao Tipo, int Para, Motivo Motivo)
{
    public static Decisao Ponto(int para, Motivo motivo) => new(TipoDeDecisao.Ponto, para, motivo);
    public static Decisao Falta(Motivo motivo) => new(TipoDeDecisao.Falta, -1, motivo);
    public static Decisao Let() => new(TipoDeDecisao.Let, -1, Motivo.Rede);
}

/// <summary>
/// Transforma os eventos da física em decisão de ponto, pelas regras do padel:
/// a bola precisa cruzar a rede e quicar no chão do outro lado ANTES de tocar qualquer parede de lá;
/// quicou do lado de quem bateu é ponto contra; bater nas próprias paredes antes de cruzar é permitido;
/// depois de quicar, paredes à vontade e o segundo quique encerra; saiu por cima: a favor de quem
/// bateu se já tinha quicado, senão contra; saque na caixa diagonal, duas faltas é ponto, let se tocou a rede
/// e caiu na caixa; quem recebe o saque não pode voleiar; o mesmo time não bate duas vezes seguidas.
/// </summary>
public sealed class Arbitro
{
    public int Faltas { get; private set; }
    public bool EmSaque { get; private set; }
    /// <summary>Time que bateu por último; null antes do saque.</summary>
    public int? Golpeador { get; private set; }
    public bool QuicouNoReceptor { get; private set; }
    public bool TocouARede { get; private set; }
    public bool Cruzou { get; private set; }
    private Caixa _caixa;

    public int? Receptor => Golpeador is int g ? 1 - g : null;

    private void ZerarRally()
    {
        EmSaque = false; Golpeador = null; QuicouNoReceptor = false; TocouARede = false; Cruzou = false;
    }

    public void NovoPonto() { Faltas = 0; ZerarRally(); }

    public void IniciarSaque(int time, Caixa caixa)
    {
        ZerarRally();
        EmSaque = true; Golpeador = time; _caixa = caixa;
    }

    public void RegistrarGolpe(int time)
    {
        ZerarRally();
        Golpeador = time;
    }

    public bool PodeGolpear(int time)
    {
        if (Golpeador is null || Golpeador == time) return false;
        if (EmSaque && !QuicouNoReceptor) return false;
        return true;
    }

    /// <summary>null = segue o jogo.</summary>
    public Decisao? Processar(EventoDaBola evento)
    {
        if (Golpeador is not int golpeador) return null;
        int ladoDoGolpeador = Quadra.LadoDoTime(golpeador);
        switch (evento.Tipo)
        {
            case TipoDeEventoDaBola.CruzouRede:
                if (evento.Para != ladoDoGolpeador) { Cruzou = true; return null; }
                return QuicouNoReceptor ? Decisao.Ponto(golpeador, Motivo.VoltouPeloVidro) : null;
            case TipoDeEventoDaBola.Rede:
                TocouARede = true;
                return null;
            case TipoDeEventoDaBola.Quique:
                if (evento.Lado == ladoDoGolpeador) return ContraOGolpeador(TocouARede ? Motivo.Rede : Motivo.NaoPassou);
                if (QuicouNoReceptor) return Decisao.Ponto(golpeador, Motivo.DoisQuiques);
                QuicouNoReceptor = true;
                if (EmSaque)
                {
                    if (!_caixa.Contem(evento.X, evento.Y)) return ContraOGolpeador(Motivo.ForaDaCaixa);
                    if (TocouARede) return Decisao.Let();
                }
                return null;
            case TipoDeEventoDaBola.Parede:
                if (evento.Lado == ladoDoGolpeador) return EmSaque ? ContraOGolpeador(Motivo.ParedePropria) : null;
                return QuicouNoReceptor ? null : ContraOGolpeador(Motivo.ParedeSemQuicar);
            case TipoDeEventoDaBola.Saiu:
                return QuicouNoReceptor ? Decisao.Ponto(golpeador, Motivo.Fora) : ContraOGolpeador(Motivo.Fora);
            default:
                return null;
        }
    }

    private Decisao ContraOGolpeador(Motivo motivo)
    {
        int receptor = Receptor!.Value;
        if (!EmSaque) return Decisao.Ponto(receptor, motivo);
        Faltas += 1;
        return Faltas >= 2 ? Decisao.Ponto(receptor, Motivo.DuplaFalta) : Decisao.Falta(motivo);
    }
}
